pallets/gear-builtin/src/tests/bls381.rs

```diff
@@ -0,0 +1,695 @@
+// This file is part of Gear.
+
+// Copyright (C) Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::mock::*;
+use ark_bls12_381::{Bls12_381, G1Affine, G1Projective as G1, G2Affine, G2Projective as G2};
+use ark_ec::{
+    pairing::Pairing,
+    short_weierstrass::{Projective as SWProjective, SWCurveConfig},
+    Group, ScalarMul, VariableBaseMSM,
+};
+use ark_ff::biginteger::BigInt;
+use ark_scale::hazmat::ArkScaleProjective;
+use ark_std::{ops::Mul, UniformRand};
+use common::Origin;
+use frame_support::assert_ok;
+use gbuiltin_bls381::*;
+use gear_core::ids::ProgramId;
+use gear_core_errors::{ErrorReplyReason, ReplyCode, SimpleExecutionError};
+use pallet_gear::GasInfo;
+use parity_scale_codec::{Decode, Encode};
+use primitive_types::H256;
+
+type ArkScale<T> = ark_scale::ArkScale<T, { ark_scale::HOST_CALL }>;
+type ScalarField = <G2 as Group>::ScalarField;
+
+const ACTOR_ID: [u8; 32] =
+    hex_literal::hex!("6b6e292c382945e80bf51af2ba7fe9f458dcff81ae6075c46f9095e1bbecdc37");
+
+pub(crate) fn init_logger() {
+    let _ = env_logger::Builder::from_default_env()
+        .format_module_path(false)
+        .format_level(true)
+        .try_init();
+}
+
+fn get_gas_info(builtin_id: ProgramId, payload: Vec<u8>) -> GasInfo {
+    start_transaction();
+    let res = Gear::calculate_gas_info(
+        SIGNER.into_origin(),
+        pallet_gear::manager::HandleKind::Handle(builtin_id),
+        payload,
+        0,
+        true,
+        None,
+        None,
+    )
+    .expect("calculate_gas_info failed");
+    rollback_transaction();
+
```

Comment:
could you pls add here assertion that res.min_gas and res.burned != 0 && != block gas limit (or < block gas limit * 90%)